### DISCUSSION TITLE
Question about DNS Caching

### DISCUSSION THREAD

**Discussion by Dynamic5912** [AdguardTeam/AdGuardHome#7428] [Q&A]
I have a question about how DNS Caching works with AGH.

- If I enable the Cache (i.e. specify a Cache Size) - what happens when items expire - are they retained or deleted from the cache?

- If I enable _Optimistic Caching_ - does this mean items are cached indefinitely (retained forever?) even after expiration and then served and refreshed in the background (i.e. like Unbound's _serve-expired_ setting)?

On the Configuration Page I see a reference to this:

_cache_time: Time interval in minutes for keeping cache records._

And in the YAML file it's set to "30" - does this mean cached items are only retained for 30 minutes?

Does this get reset each time they are called from the cache and updated?

**Comment by schzhn** [comment]
Cached responses are retained indefinitely until they are replaced by new responses. This depends on the size of the cache (`dns.cache_size`) and the number of stored responses.

If you enable the "Optimistic Caching" setting (`dns.cache_optimistic`), expired responses are still provided with a Time-To-Live (TTL) of 10.  In the background, AdGuard Home queries the upstream server and updates the cached response, so subsequent requests receive the updated response.
See https://github.com/AdguardTeam/dnsproxy/blob/129cf05264e1945ccbb2d1b9f8a5fe62522965d3/proxy/cache.go#L115

The "filtering.cache_time" setting defines how long hashes, used for safe browsing and parental control, are stored.

**Reply by Dynamic5912** [reply]
> Cached responses are retained indefinitely until they are replaced by new responses. This depends on the size of the cache (`dns.cache_size`) and the number of stored responses.
> 
> 
> 
> If you enable the "Optimistic Caching" setting (`dns.cache_optimistic`), expired responses are still provided with a Time-To-Live (TTL) of 10.  In the background, AdGuard Home queries the upstream server and updates the cached response, so subsequent requests receive the updated response.
> 
> See https://github.com/AdguardTeam/dnsproxy/blob/129cf05264e1945ccbb2d1b9f8a5fe62522965d3/proxy/cache.go#L115
> 
> 
> 
> The "filtering.cache_time" setting defines how long hashes, used for safe browsing and parental control, are stored.
> 
> 

Thanks. 

So when a cached record expires (I.e the TTL runs out) it's still retained in the cache and served to clients?

Or is it deleted and replaced with the updated record when a request is made?

I understand optimising caching does this in the background when enabled - but if it's **not** enabled how do cached records behave?

**Reply by schzhn** [reply]
In the cache, AdGuardHome stores the last N responses. When it receives a request, it checks the cache to see if there is a non-expired response available. If so, it serves that response. Otherwise, it queries the upstream server, serves the response from the upstream, and updates the cache.

If optimistic caching is enabled, AdGuardHome serves an expired response with a short TTL and updates the cache in the background.

**Reply by ngocphamm** [reply]
In my instance at home, those domains are being queried every minute and I have never seen them being served from cache. Does anyone has some sort of explanation?

Same domains, coming from same client.

![CleanShot 2025-01-24 at 07 03 44@2x](https://github.com/user-attachments/assets/fd6bcac0-c868-4ac2-9050-14b419b3dde2)

**Reply by Dynamic5912** [reply]
Are there custom settings for the client?

If so, have you enabled the caching for that client within the settings?

**Reply by ngocphamm** [reply]
@Dynamic5912 

What kind of custom settings are you referring to? If within AdGuard, no I don't think so.

This is my DNS cache settings of AdGuard. I'm not sure how to enable caching for one client only or so.

![CleanShot 2025-01-24 at 07 25 10@2x](https://github.com/user-attachments/assets/a1a7912a-f5f7-4382-bdcf-7613244bab1d)

**Reply by Dynamic5912** [reply]
Check the client settings for the client in question and check if you're using a custom upstream DNS server.

And if you are, check if the cache configuration is enabled within that same settings page/section.

**Reply by ngocphamm** [reply]
@Dynamic5912 If you mean this page, here it is. I don't have any custom settings for any of the clients.

![CleanShot 2025-01-24 at 10 34 50@2x](https://github.com/user-attachments/assets/75e0ed52-aa25-4dae-b02c-31f75dc6f733)

**Reply by schzhn** [reply]
@ngocphamm, in your case, the TTL (Time to Live) of the DNS record is set to one minute or less. Because of this short lifespan, the records are not served from the cache.
You can confirm this by using tools like `dig` or `nslookup`.

**Reply by ngocphamm** [reply]
@schzhn I see. Thanks for pointing that out.

There is the option in DNS caching to override min TTL, but is there any way I can override TTL for those specific domains, so it will be cached for longer time? I tried to look at https://adguard-dns.io/kb/general/dns-filtering-syntax/ but I couldn't see anything related to TTL.

**Reply by schzhn** [reply]
@ngocphamm, currently, there is no way to set a custom TTL for specific DNS rewrites.
Related #1518
